visual_behavior/dimensionality_reduction/clustering/plotting.py

```diff
@@ -1,29 +1,152 @@
+
 from scipy import signal
+# from scipy.stats import spearmanr
+# from scipy.stats import kruskal
+# from scipy.stats import ttest_ind
+from scipy.spatial.distance import cdist, pdist
+
+from sklearn.cluster import KMeans
+# from sklearn.metrics import silhouette_score
+# from sklearn.metrics import pairwise_distances
+# from sklearn.cluster import SpectralClustering
+
 import visual_behavior.visualization.utils as utils
+import visual_behavior.data_access.loading as loading
+import visual_behavior.data_access.utilities as utilities
+# import visual_behavior_glm.GLM_analysis_tools as gat
+
+from visual_behavior.dimensionality_reduction.clustering import processing
+from visual_behavior.dimensionality_reduction.clustering.processing import get_silhouette_scores, get_cluster_density, get_cre_lines, get_cell_type_for_cre_line
+
+import seaborn as sns
+sns.set_context('notebook', font_scale=1.5, rc={'lines.markeredgewidth': 2})
+
+
+def plot_feature_matrix_for_cre_lines(feature_matrix, cell_metadata, save_dir=None, folder=None):
+    """
+    plots the feature matrix used for clustering where feature matrix consists of cell_specimen_ids as rows,
+    and features x experience levels as columns, for cells matched across experience levels
+    :param feature_matrix:
+    :param cell_metadata: table with cell_specimen_id as index and metadata as columns
+    :param save_dir: directory to save plot
+    :param folder:
+    :return:
+    """
+    # check if there are negative values in feature_matrix, if so, use diff cmap and set vmin to -1
+    if len(np.where(feature_matrix < 0)[0]) > 0:
+        vmin = -1
+        cmap = 'RdBu'
+    else:
+        vmin = 0
+        cmap = 'Blues'
+
+    figsize = (15,5)
+    fig, ax = plt.subplots(1,3, figsize=figsize)
+    for i, cre_line in enumerate(get_cre_lines(cell_metadata)):
+        data = processing.get_feature_matrix_for_cre_line(feature_matrix, cell_metadata, cre_line)
+        ax[i] = sns.heatmap(data.values, cmap=cmap, ax=ax[i], vmin=vmin, vmax=1,
+                         robust=True, cbar_kws={"drawedges": False, "shrink": 0.7, "label": 'coding score'})
+        ax[i].set_title(get_cell_type_for_cre_line(cell_metadata, cre_line))
+        ax[i].set_ylabel('cells')
+        ax[i].set_ylim(0, data.shape[0])
+        ax[i].set_yticks([0, data.shape[0]]);
+        ax[i].set_yticklabels((0, data.shape[0]), fontsize=14);
+        ax[i].set_ylim(ax[i].get_ylim()[::-1]) #flip y axes so larger clusters are on top
+        ax[i].set_xlabel('')
+        ax[i].set_xlim(0, data.shape[1])
+        ax[i].set_xticks(np.arange(0, data.shape[1])+0.5)
+        ax[i].set_xticklabels([key[1]+' -  '+key[0] for key in list(data.keys())], rotation=90, fontsize=14);
+
+    fig.subplots_adjust(wspace=0.7)
+
+    if save_dir:
+        utils.save_figure(fig, figsize, save_dir, folder, 'feature_matrix_unsorted')
+
+
+def plot_feature_matrix_sorted(feature_matrix, cluster_meta, sort_col='cluster_id', save_dir=None, folder=None):
```

Comment:
can this function and above function be converged with an input variable of "to_sort" or something like that?